shared/src/main/java/org/geysermc/hydraulic/pack/PackManager.java

```diff
@@ -157,4 +210,93 @@ private void callEvent(@NotNull ModInfo mod, @NotNull Event event) {
             module.call(event.getClass(), new PackEventContext(this.hydraulic, mod, module, event));
         }
     }
+
+    private void initializeModLookups() {
+        final Multimap<String, ModInfo> namespacesToMods = this.namespacesToMods;
+        namespacesToMods.clear();
+        for (final ModInfo mod : hydraulic.mods()) {
+            for (final Path root : mod.roots()) {
+                final Path assets = root.resolve("assets");
+                if (!Files.isDirectory(assets)) continue;
+                try (Stream<Path> stream = Files.list(assets)) {
+                    IOStream.adapt(stream)
+                        .filter(Files::isDirectory)
+                        .unwrap()
+                        .map(Path::getFileName)
+                        .map(Path::toString)
+                        .filter(namespace -> !namespace.equals("minecraft"))
+                        .forEach(namespace -> namespacesToMods.put(namespace, mod));
+                } catch (IOException e) {
+                    LOGGER.error("Failed to list namespaces for mod {}", mod.id(), e);
+                }
+            }
+        }
+
+        final Multimap<String, ResourceLocation> modsToBlocks = this.modsToBlocks;
+        modsToBlocks.clear();
+        blocksLoop:
+        for (final ResourceLocation block : BuiltInRegistries.BLOCK.keySet()) {
+            if (block.getNamespace().equals("minecraft")) continue;
+            for (final ModInfo mod : namespacesToMods.get(block.getNamespace())) {
+                final Path checkFile = mod.resolveFile("assets/" + block.getNamespace() + "/blockstates/" + block.getPath() + ".json");
+                if (checkFile != null) {
+                    modsToBlocks.put(mod.id(), block);
+                    continue blocksLoop;
+                }
+            }
+        }
+
+        final Multimap<String, ResourceLocation> modsToItems = this.modsToItems;
+        modsToItems.clear();
+        itemsLoop:
+        for (final ResourceLocation item : BuiltInRegistries.ITEM.keySet()) {
+            if (item.getNamespace().equals("minecraft")) continue;
+            for (final ModInfo mod : namespacesToMods.get(item.getNamespace())) {
+                final Path checkFile = mod.resolveFile("assets/" + item.getNamespace() + "/models/item/" + item.getPath() + ".json");
+                if (checkFile != null) {
+                    modsToItems.put(mod.id(), item);
+                    continue itemsLoop;
+                }
+            }
+        }
+    }
+
+    // Based off of ModelStitcher.vanillaProvider
```

Comment:
Can we expand this to a jdoc and explain whats different/why we cant use vanillaProvider?